crypto/dso/dso_dlfcn.c

```diff
@@ -359,14 +363,27 @@ static int dladdr(void *ptr, Dl_info *dl)
             || ((addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
                 && (addr < ((uintptr_t)this_ldi->ldinfo_dataorg +
                             this_ldi->ldinfo_datasize)))) {
+            char *fname;
+            char *member = (char *)((uintptr_t)this_ldi->ldinfo_filename +
+                                    strlen(this_ldi->ldinfo_filename) + 1);
+            size_t member_len = strlen(member);
             found = 1;
-            /*
-             * Ignoring the possibility of a member name and just returning
-             * the path name. See docs: sys/ldr.h, loadquery() and
-             * dlopen()/RTLD_MEMBER.
-             */
-            if ((dl->dli_fname =
-                 OPENSSL_strdup(this_ldi->ldinfo_filename)) == NULL)
+            if ((fname = OPENSSL_strdup(this_ldi->ldinfo_filename)) != NULL) {
+                if ((member_len > 0)
+                    && (fname = OPENSSL_realloc((void *)fname, strlen(fname) +
+                                                    1 + member_len + 2))) {
```

Comment:
Why bother allocating and reallocating, when you can allocate what's needed at once?

BTW, while you are at it, line 335 is too long and needs to be broken.